compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs

```diff
@@ -1100,20 +1207,56 @@ impl Context {
                 .into());
             }
         }
+
         // The final array should will the flattened index at each outer array index
         let init_values = vecmap(flat_elem_type_sizes, |type_size| {
             let var = self.acir_context.add_constant(FieldElement::from(type_size as u128));
             AcirValue::Var(var, AcirType::field())
         });
+        let element_type_sizes_len = init_values.len();
         self.initialize_array(
             element_type_sizes,
-            init_values.len(),
+            element_type_sizes_len,
             Some(AcirValue::Array(init_values.into())),
         )?;
 
+        self.internal_mem_block_lengths.insert(element_type_sizes, element_type_sizes_len);
+
         Ok(element_type_sizes)
     }
 
+    fn compute_slice_sizes(
+        &mut self,
+        current_array_id: ValueId,
+        parent_array: Option<ValueId>,
+        dfg: &DataFlowGraph,
+    ) {
+        if let Value::Array { array, typ } = &dfg[current_array_id] {
+            if let Type::Slice(_) = typ {
+                let element_size = typ.element_size();
+                let true_len = array.len() / element_size;
+                if let Some(parent_array) = parent_array {
+                    let sizes_list =
+                        self.slice_sizes.get_mut(&parent_array).expect("ICE: expected size list");
+                    sizes_list.push(true_len);
+                } else {
+                    // This means the current_array_id is the parent array
+                    self.slice_sizes.insert(current_array_id, vec![true_len]);
+                }
+                for value in array {
+                    let typ = dfg.type_of_value(*value);
+                    if let Type::Slice(_) = typ {
+                        if parent_array.is_some() {
+                            self.compute_slice_sizes(*value, parent_array, dfg);
+                        } else {
+                            self.compute_slice_sizes(*value, Some(current_array_id), dfg);
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
We can remove two layers of nesting here with an early return.

Reply:
done